Not working: parent.join(child.join(grandchild)) [in some cases]

## Setup

### Versions

- **Rust:** rustc 1.21.0-nightly (b75d1f0ce 2017-08-02)
- **Diesel:** 0.15.2
- **Database:** PG
- **Operating System** macOS 10.12.6

### Feature Flags

- **diesel:** ["postgres"]
- **diesel_codegen:** ["postgres"]

## Problem Description

In certain relational schemas, performing a `parent.join(child.join(grandchild))` doesn't work as expected (`join` can be an `inner_join` or `left_join`, same error).

The same code, using the simpler `parent.join(child.join)` call works.

**If I change the schema and reverse the belongs_to relation between parent and child, the nested join works just fine. However that's the schema I want to rely on, plus my schema is SQL-valid.**

### What are you trying to accomplish?

I'm trying to perform a join w/ the following structure: `parent.join(child.join(grandchild))`

Given the following schema: 

```rust
table! {
    parent (id) {
        id -> Integer,
        child_id -> Integer,
    }
}

table! {
    child (id) {
        id -> Integer,
        grand_child_id -> Integer,
    }
}

table! {
    grandchild (id) {
        id -> Integer,
    }
}
```

And the following model:

```rust
#[derive(Identifiable, Queryable, Associations)]
#[belongs_to(Child)]
#[table_name="parent"]
pub struct Parent {
    pub id: i32,
    pub child_id: i32,
}

#[derive(Identifiable, Queryable, Associations)]
#[belongs_to(GrandChild)]
#[table_name="child"]
pub struct Child {
    pub id: i32,
    pub grand_child_id: i32,
}

#[derive(Identifiable, Queryable, Associations)]
#[table_name="grandchild"]
pub struct GrandChild {
    pub id: i32,
}
```

### What is the expected output?

The expected output is the equivalent of the full-SQL request:

```sql
SELECT "person_employment"."title", "company"."legal_name", "contact"."emails" FROM (
	"person_employment"
		INNER JOIN "company" ON "person_employment"."company_id" = "company"."id"
		INNER JOIN "contact" ON "company"."contact_id" = "contact"."id"
);
```

Which works just fine in a PG prompt, based on the bare SQL schema defined above.

### What is the actual output?

Getting the following error when compiling, although all relevant `belongs_to` and `enable_multi_table_joins!` are set.

```
error[E0277]: the trait bound `diesel::query_source::joins::PleaseGenerateInverseJoinImpls<parent::table>: diesel::JoinTo<child::table>` is not satisfied
   --> src/lib.rs:141:10
    |
141 |         .inner_join(child::table.inner_join(grandchild::table))
    |          ^^^^^^^^^^ the trait `diesel::JoinTo<child::table>` is not implemented for `diesel::query_source::joins::PleaseGenerateInverseJoinImpls<parent::table>`
```

### Are you seeing any additional errors?

No, only this one.

### Steps to reproduce

1. Gist: [valeriansaliou/48f441bbcdcef29f5395b87b09dc9893](https://gist.github.com/valeriansaliou/48f441bbcdcef29f5395b87b09dc9893)
2. Compile the code w/o changing it. It works. It's using the only relational schema I could get it to work with, however I don't want it this way (in this schema, a child references both its parent and grandchild)
3. Uncomment the 2 commented sections regarding schema + models and comment out the other ones.
4. Build it, you'll get the compilation error I'm talking about. This is using the schema that's more appropriate in that case, but which will never build w/ Diesel.

Yeah it's backwards, because otherwise I get:

```
error[E0275]: overflow evaluating the requirement `<child::table as diesel::JoinTo<grandchild::table>>::OnClause`
   --> src/lib.rs:141:34
    |
141 |         .inner_join(child::table.inner_join(grandchild::table))
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -26,6 +26,9 @@ for Rust libraries in [RFC #1105](https://github.com/rust-lang/rfcs/blob/master/
   backend specific SQL. (The specific backend they will generate for will be
   arbitrarily chosen based on the backends enabled).
 
+* `#[belongs_to]` will no longer generate the code required to join between two
+  tables. You will need to explicitly invoke `joinable!` instead.
+
 ### Removed
 
 * `debug_sql!` has been deprecated in favor of `diesel::debug_sql`.
@@ -33,6 +36,12 @@ for Rust libraries in [RFC #1105](https://github.com/rust-lang/rfcs/blob/master/
 * `print_sql!` has been deprecated without replacement.
 
 * `diesel::backend::Debug` has been removed.
+
+### Fixed
+
+* Diesel now properly supports joins in the form:
+  `grandchild.join(child.join(parent))`. Previously only
+  `parent.join(child.join(grandchild))` would compile.
 
 [bigdecimal-0.14.0]: https://crates.io/crates/bigdecimal
 

--- a/diesel/src/macros/associations/belongs_to.rs
+++ b/diesel/src/macros/associations/belongs_to.rs
@@ -159,17 +159,6 @@ macro_rules! BelongsTo {
                 $child_table_name::$foreign_key_name
             }
         }
-
-        BelongsTo! {
-            @generate_joins,
-            (
-                struct_name = $struct_name,
-                parent_struct = $parent_struct,
-                foreign_key_name = $foreign_key_name,
-                child_table_name = $child_table_name,
-            ),
-            $($rest)*
-        }
     };
 
     // Generate code when FK is optional
@@ -196,54 +185,6 @@ macro_rules! BelongsTo {
                 $child_table_name::$foreign_key_name
             }
         }
-
-        BelongsTo! {
-            @generate_joins,
-            (
-                struct_name = $struct_name,
-                parent_struct = $parent_struct,
-                foreign_key_name = $foreign_key_name,
-                child_table_name = $child_table_name,
-            ),
-            $($rest)*
-        }
-    };
-
-    // Generate code that does not differ based on the fk being optional
-    (
-        @generate_joins,
-        (
-            struct_name = $struct_name:ident,
-            parent_struct = $parent_struct:ident,
-            foreign_key_name = $foreign_key_name:ident,
-            child_table_name = $child_table_name:ident,
-        ),
-        fields = [$({
-            field_name: $field_name:ident,
-            column_name: $column_name:ident,
-            field_ty: $field_ty:ty,
-            field_kind: $field_kind:ident,
-            $($rest:tt)*
-        })+],
-    ) => {
-        static_cond!(if $struct_name != $parent_struct {
-            joinable_inner!(
-                left_table_ty = $child_table_name::table,
-                right_table_ty = <$parent_struct as $crate::associations::HasTable>::Table,
-                right_table_expr = <$parent_struct as $crate::associations::HasTable>::table(),
-                foreign_key = $child_table_name::$foreign_key_name,
-                primary_key_ty = <<$parent_struct as $crate::associations::HasTable>::Table as $crate::Table>::PrimaryKey,
-                primary_key_expr = $crate::Table::primary_key(&<$parent_struct as $crate::associations::HasTable>::table()),
-            );
-            joinable_inner!(
-                left_table_ty = $child_table_name::table,
-                right_table_ty = $crate::query_source::joins::PleaseGenerateInverseJoinImpls<<$parent_struct as $crate::associations::HasTable>::Table>,
-                right_table_expr = <$parent_struct as $crate::associations::HasTable>::table(),
-                foreign_key = $child_table_name::$foreign_key_name,
-                primary_key_ty = <<$parent_struct as $crate::associations::HasTable>::Table as $crate::Table>::PrimaryKey,
-                primary_key_expr = $crate::Table::primary_key(&<$parent_struct as $crate::associations::HasTable>::table()),
-            );
-        });
     };
 
     // Handle struct with no generics

--- a/diesel/src/macros/mod.rs
+++ b/diesel/src/macros/mod.rs
@@ -563,7 +563,7 @@ macro_rules! table_body {
             use $crate::query_builder::*;
             use $crate::query_builder::nodes::Identifier;
             use $crate::query_source::{AppearsInFromClause, Once, Never};
-            use $crate::query_source::joins::PleaseGenerateInverseJoinImpls;
+            use $crate::query_source::joins::{Join, JoinOn};
             $(use $($import)::+;)+
             pub use self::columns::*;
 
@@ -653,14 +653,49 @@ macro_rules! table_body {
                 type Count = Never;
             }
 
-            impl<T> JoinTo<T> for table where
-                T: JoinTo<table> + JoinTo<PleaseGenerateInverseJoinImpls<table>>,
+            impl<Left, Right, Kind> JoinTo<Join<Left, Right, Kind>> for table where
+                Join<Left, Right, Kind>: JoinTo<table>,
             {
-                type FromClause = T;
-                type OnClause = <T as JoinTo<table>>::OnClause;
+                type FromClause = Join<Left, Right, Kind>;
+                type OnClause = <Join<Left, Right, Kind> as JoinTo<table>>::OnClause;
 
-                fn join_target(rhs: T) -> (Self::FromClause, Self::OnClause) {
-                    let (_, on_clause) = T::join_target(table);
+                fn join_target(rhs: Join<Left, Right, Kind>) -> (Self::FromClause, Self::OnClause) {
+                    let (_, on_clause) = Join::join_target(table);
+                    (rhs, on_clause)
+                }
+            }
+
+            impl<Join, On> JoinTo<JoinOn<Join, On>> for table where
+                JoinOn<Join, On>: JoinTo<table>,
+            {
+                type FromClause = JoinOn<Join, On>;
+                type OnClause = <JoinOn<Join, On> as JoinTo<table>>::OnClause;
+
+                fn join_target(rhs: JoinOn<Join, On>) -> (Self::FromClause, Self::OnClause) {
+                    let (_, on_clause) = JoinOn::join_target(table);
+                    (rhs, on_clause)
+                }
+            }
+
+            impl<F, S, D, W, O, L, Of, G> JoinTo<SelectStatement<F, S, D, W, O, L, Of, G>> for table where
+                SelectStatement<F, S, D, W, O, L, Of, G>: JoinTo<table>,
+            {
+                type FromClause = SelectStatement<F, S, D, W, O, L, Of, G>;
+                type OnClause = <SelectStatement<F, S, D, W, O, L, Of, G> as JoinTo<table>>::OnClause;
+
+                fn join_target(rhs: SelectStatement<F, S, D, W, O, L, Of, G>) -> (Self::FromClause, Self::OnClause) {
+                    let (_, on_clause) = SelectStatement::join_target(table);
+                    (rhs, on_clause)
+                }
+            }
+
+            impl<'a, QS, ST, DB> JoinTo<BoxedSelectStatement<'a, QS, ST, DB>> for table where
+                BoxedSelectStatement<'a, QS, ST, DB>: JoinTo<table>,
+            {
+                type FromClause = BoxedSelectStatement<'a, QS, ST, DB>;
+                type OnClause = <BoxedSelectStatement<'a, QS, ST, DB> as JoinTo<table>>::OnClause;
+                fn join_target(rhs: BoxedSelectStatement<'a, QS, ST, DB>) -> (Self::FromClause, Self::OnClause) {
+                    let (_, on_clause) = BoxedSelectStatement::join_target(table);
                     (rhs, on_clause)
                 }
             }

--- a/diesel/src/query_source/joins.rs
+++ b/diesel/src/query_source/joins.rs
@@ -266,22 +266,6 @@ impl<T, Join, On> AppearsInFromClause<T> for JoinOn<Join, On> where
     type Count = Join::Count;
 }
 
-#[allow(missing_debug_implementations, missing_copy_implementations)]
-#[doc(hidden)]
-/// A hack to allow bidirectional joins to be generated from `#[belongs_to]`
-///
-/// This type needs to exist because it is illegal in Rust today to write
-/// `impl JoinTo<posts> for <User as HasTable>::Table`, even though the type
-/// is fully monomorphic and projects to a local type. If this restriction
-/// were ever lifted in the future, this type could be removed.
-///
-/// Instead, after generating `impl JoinTo<<User as HasTable>::Table> for
-/// posts`, we *also* generate `impl JoinTo<PleaseGenerateInverseJoinImpls<<User
-/// as HasTable>::table> for posts`, and rely on the fact that `users::table`
-/// will have a blanket impl on itself for anything that joins to
-/// `PleaseGenerateInverseJoinImpls`.
-pub struct PleaseGenerateInverseJoinImpls<T>(T);
-
 pub trait Plus<T> {
     type Output;
 }

--- a/diesel_compile_tests/tests/compile-fail/aggregate_expression_requires_column_from_same_table.rs
+++ b/diesel_compile_tests/tests/compile-fail/aggregate_expression_requires_column_from_same_table.rs
@@ -21,28 +21,16 @@ fn main() {
     //~^ ERROR E0277
     //~| ERROR AppearsInFromClause
     //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
     let source = users::table.select(avg(posts::id));
     //~^ ERROR E0277
     //~| ERROR AppearsInFromClause
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
     //~| ERROR E0277
     let source = users::table.select(max(posts::id));
     //~^ ERROR E0277
     //~| ERROR AppearsInFromClause
     //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
     let source = users::table.select(min(posts::id));
     //~^ ERROR E0277
     //~| ERROR AppearsInFromClause
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
     //~| ERROR E0277
 }

--- a/diesel_compile_tests/tests/compile-fail/any_is_only_selectable_if_inner_expr_is_selectable.rs
+++ b/diesel_compile_tests/tests/compile-fail/any_is_only_selectable_if_inner_expr_is_selectable.rs
@@ -22,8 +22,5 @@ fn main() {
 
     stuff.filter(name.eq(any(more_stuff::names)));
     //~^ ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
     //~| ERROR E0271
 }

--- a/diesel_compile_tests/tests/compile-fail/boxed_queries_require_selectable_expression_for_filter.rs
+++ b/diesel_compile_tests/tests/compile-fail/boxed_queries_require_selectable_expression_for_filter.rs
@@ -22,7 +22,4 @@ fn main() {
     users::table.into_boxed::<Pg>().filter(posts::title.eq("Hello"));
     //~^ ERROR AppearsInFromClause
     //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
 }

--- a/diesel_compile_tests/tests/compile-fail/boxed_queries_require_selectable_expression_for_order.rs
+++ b/diesel_compile_tests/tests/compile-fail/boxed_queries_require_selectable_expression_for_order.rs
@@ -22,7 +22,4 @@ fn main() {
     users::table.into_boxed::<Pg>().order(posts::title.desc());
     //~^ ERROR AppearsInFromClause
     //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
 }

--- a/diesel_compile_tests/tests/compile-fail/cannot_join_to_non_joinable_table.rs
+++ b/diesel_compile_tests/tests/compile-fail/cannot_join_to_non_joinable_table.rs
@@ -30,13 +30,13 @@ enable_multi_table_joins!(users, comments);
 
 fn main() {
     let _ = users::table.inner_join(posts::table);
-    //~^ ERROR 0275
+    //~^ ERROR 0277
     let _ = users::table.left_outer_join(posts::table);
-    // We would get an error here but 0275 halts everything
+    //~^ ERROR 0277
 
     // Sanity check to make sure the error is when users
     // become involved
     let join = posts::table.inner_join(comments::table);
     let _ = users::table.inner_join(join);
-    // We would get an error here but 0275 halts everything
+    //~^ ERROR 0277
 }

--- a/diesel_compile_tests/tests/compile-fail/custom_returning_requires_selectable_expression.rs
+++ b/diesel_compile_tests/tests/compile-fail/custom_returning_requires_selectable_expression.rs
@@ -38,7 +38,4 @@ fn main() {
     //~^ ERROR SelectableExpression
     //~| ERROR AppearsInFromClause
     //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
 }

--- a/diesel_compile_tests/tests/compile-fail/filter_cannot_take_comparison_for_columns_from_another_table.rs
+++ b/diesel_compile_tests/tests/compile-fail/filter_cannot_take_comparison_for_columns_from_another_table.rs
@@ -21,13 +21,7 @@ fn main() {
     let _ = users::table.filter(posts::id.eq(1));
     //~^ ERROR AppearsInFromClause
     //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
     let _ = users::table.filter(users::name.eq(posts::title));
     //~^ ERROR AppearsInFromClause
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
     //~| ERROR E0277
 }

--- a/diesel_compile_tests/tests/compile-fail/order_requires_column_from_same_table.rs
+++ b/diesel_compile_tests/tests/compile-fail/order_requires_column_from_same_table.rs
@@ -18,8 +18,5 @@ table! {
 fn main() {
     let source = users::table.order(posts::id);
     //~^ ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
     //~| ERROR E0271
 }

--- a/diesel_compile_tests/tests/compile-fail/pg_upsert_do_update_requires_valid_update.rs
+++ b/diesel_compile_tests/tests/compile-fail/pg_upsert_do_update_requires_valid_update.rs
@@ -41,9 +41,6 @@ fn main() {
     insert(&NewUser("Sean").on_conflict(id, do_update().set(name.eq(posts::title)))).into(users).execute(&connection);
     //~^ ERROR E0277
     //~| ERROR no method named `execute`
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
     //~| ERROR E0271
 
     // Update column with excluded value that is not selectable

--- a/diesel_compile_tests/tests/compile-fail/selecting_multiple_columns_requires_all_must_be_from_selectable_table.rs
+++ b/diesel_compile_tests/tests/compile-fail/selecting_multiple_columns_requires_all_must_be_from_selectable_table.rs
@@ -23,15 +23,9 @@ fn main() {
     //~^ ERROR Selectable
     //~| ERROR E0277
     //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
     //~| ERROR E0271
     let stuff = users::table.select((posts::id, users::name));
     //~^ ERROR Selectable
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
     //~| ERROR E0277
     //~| ERROR E0271
 }

--- a/diesel_compile_tests/tests/compile-fail/update_requires_column_be_from_same_table.rs
+++ b/diesel_compile_tests/tests/compile-fail/update_requires_column_be_from_same_table.rs
@@ -24,8 +24,5 @@ fn main() {
     //~^ ERROR type mismatch
     let command = update(users).set(name.eq(posts::title));
     //~^ ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
     //~| ERROR E0271
 }

--- a/diesel_compile_tests/tests/compile-fail/user_defined_functions_follow_same_selection_rules.rs
+++ b/diesel_compile_tests/tests/compile-fail/user_defined_functions_follow_same_selection_rules.rs
@@ -29,8 +29,5 @@ fn main() {
     //~^ ERROR type mismatch
     let _ = users::table.filter(name.eq(bar(title)));
     //~^ ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
-    //~| ERROR E0277
     //~| ERROR E0271
 }

--- a/diesel_tests/tests/annotations.rs
+++ b/diesel_tests/tests/annotations.rs
@@ -14,8 +14,9 @@ fn association_where_struct_name_doesnt_match_table_name() {
     let connection = connection_with_sean_and_tess_in_users_table();
 
     let sean = find_user_by_name("Sean", &connection);
-    let post: Post = insert(&sean.new_post("Hello", None)).into(posts::table)
-        .get_result(&connection).unwrap();
+    insert(&sean.new_post("Hello", None)).into(posts::table)
+        .execute(&connection).unwrap();
+    let post = posts::table.first::<Post>(&connection).unwrap();
     insert(&NewComment(post.id, "comment")).into(comments::table)
         .execute(&connection).unwrap();
 

--- a/diesel_tests/tests/annotations.rs
+++ b/diesel_tests/tests/annotations.rs
@@ -1,30 +1,28 @@
 use diesel::*;
 use schema::*;
 
-// FIXME: Bring this test back once we can figure out how to allow multiple structs
-// on the same table to use `#[belongs_to]` without overlapping the `JoinTo` impls
-// #[test]
-// fn association_where_struct_name_doesnt_match_table_name() {
-//     #[derive(PartialEq, Eq, Debug, Clone, Queryable, Identifiable, Associations)]
-//     #[belongs_to(Post)]
-//     #[table_name="comments"]
-//     struct OtherComment {
-//         id: i32,
-//         post_id: i32
-//     }
+#[test]
+fn association_where_struct_name_doesnt_match_table_name() {
+    #[derive(PartialEq, Eq, Debug, Clone, Queryable, Identifiable, Associations)]
+    #[belongs_to(Post)]
+    #[table_name="comments"]
+    struct OtherComment {
+        id: i32,
+        post_id: i32
+    }
 
-//     let connection = connection_with_sean_and_tess_in_users_table();
+    let connection = connection_with_sean_and_tess_in_users_table();
 
-//     let sean = find_user_by_name("Sean", &connection);
-//     let post: Post = insert(&sean.new_post("Hello", None)).into(posts::table)
-//         .get_result(&connection).unwrap();
-//     insert(&NewComment(post.id, "comment")).into(comments::table)
-//         .execute(&connection).unwrap();
+    let sean = find_user_by_name("Sean", &connection);
+    let post: Post = insert(&sean.new_post("Hello", None)).into(posts::table)
+        .get_result(&connection).unwrap();
+    insert(&NewComment(post.id, "comment")).into(comments::table)
+        .execute(&connection).unwrap();
 
-//     let comment_text = OtherComment::belonging_to(&post).select(comments::text)
-//         .first::<String>(&connection);
-//     assert_eq!(Ok("comment".into()), comment_text);
-// }
+    let comment_text = OtherComment::belonging_to(&post).select(comments::text)
+        .first::<String>(&connection);
+    assert_eq!(Ok("comment".into()), comment_text);
+}
 
 #[test]
 #[cfg(not(any(feature="sqlite", feature="mysql")))]

--- a/diesel_tests/tests/joins.rs
+++ b/diesel_tests/tests/joins.rs
@@ -373,6 +373,22 @@ fn selecting_parent_child_grandchild() {
 }
 
 #[test]
+fn selecting_grandchild_child_parent() {
+    let (connection, test_data) = connection_with_fixture_data_for_multitable_joins();
+    let TestData { sean, posts, comments, .. } = test_data;
+
+    let data = comments::table.inner_join(posts::table.inner_join(users::table))
+        .order((users::id, posts::id, comments::id))
+        .load(&connection);
+    let expected = vec![
+        (comments[0].clone(), (posts[0].clone(), sean.clone())),
+        (comments[2].clone(), (posts[0].clone(), sean.clone())),
+        (comments[1].clone(), (posts[2].clone(), sean.clone())),
+    ];
+    assert_eq!(Ok(expected), data);
+}
+
+#[test]
 fn selecting_four_tables_deep() {
     let (connection, test_data) = connection_with_fixture_data_for_multitable_joins();
     let TestData { sean, posts, comments, likes, .. } = test_data;

--- a/diesel_tests/tests/schema.rs
+++ b/diesel_tests/tests/schema.rs
@@ -11,6 +11,13 @@ infer_schema!("dotenv:MYSQL_DATABASE_URL");
 #[cfg(not(feature="backend_specific_database_url"))]
 infer_schema!("dotenv:DATABASE_URL");
 
+joinable!(posts -> users (user_id));
+joinable!(comments -> posts (post_id));
+joinable!(followings -> users (user_id));
+joinable!(followings -> posts (post_id));
+joinable!(likes -> comments (comment_id));
+joinable!(likes -> users (user_id));
+
 #[derive(PartialEq, Eq, Debug, Clone, Queryable, Identifiable, Insertable, AsChangeset, Associations)]
 #[table_name = "users"]
 pub struct User {

